### ISSUE TITLE
Accidental loss of precision in calc_variance

### ISSUE THREAD

**Issue by danluu** [stan-dev/math#814]
#### Summary:

`calc_variance` has a computation which does an `integer` (`size_t`) divide and then converts the result to `double`.

#### Description:

~~~
double two_over_size_m1 = 2 / (size - 1);
  for (size_t i = 0; i < size; ++i)
    partials[i] = two_over_size_m1 * (dtrs[i].vi_->val_ - mean);
~~~

The RHS of the first line above is all integer, so the computation is done on integers and then converted to floats. For a concrete example of a case where this would make a difference, if `size` = 4, `two_over_size_m1` will be `0`, but if the calculation were done on doubles, the result would be `0.66...`. Is this on purpose?

An independent thing that seems non-optimal is that there are two loops that loop over the same thing, and I think that most compilers won't do loop fusion to merge the two loops into a single loop with "normal" optimization settings (`-O2` or `-O3`).

The entire function is:

~~~
inline var calc_variance(size_t size, const var* dtrs) {
  vari** varis = reinterpret_cast<vari**>(
      ChainableStack::memalloc_.alloc(size * sizeof(vari*)));
  for (size_t i = 0; i < size; ++i)
    varis[i] = dtrs[i].vi_;
  double sum = 0.0;
  for (size_t i = 0; i < size; ++i)
    sum += dtrs[i].vi_->val_;
  double mean = sum / size;
  double sum_of_squares = 0;
  for (size_t i = 0; i < size; ++i) {
    double diff = dtrs[i].vi_->val_ - mean;
    sum_of_squares += diff * diff;
  }
  double variance = sum_of_squares / (size - 1);
  double* partials = reinterpret_cast<double*>(
      ChainableStack::memalloc_.alloc(size * sizeof(double)));
  double two_over_size_m1 = 2 / (size - 1);
  for (size_t i = 0; i < size; ++i)
    partials[i] = two_over_size_m1 * (dtrs[i].vi_->val_ - mean);
  return var(new stored_gradient_vari(variance, size, varis, partials));
}
~~~

To compute both `partials` and `variance`, we loop over `dtrs` and compute `dtrs[i].vi_->val_ - mean`, but in two separate loops.

[Here is an example of some much simpler code, with assembly output, where I believe both clang and gcc fail to merge the two loops, resulting in twice as many loads as necessary](https://godbolt.org/g/vMmsXa).

#### Current Version:
v2.17.0

**Comment by bob-carpenter** [comment]
That's definitely a bug---replacing `2` with `2.0` in the numerator will fix it.

It's also the naive algorithm for computing variance in that there can be a catastrophic loss of precision in calculating `diff`.  Welford'a algorithm is much more stable.  We should have just a good sample variance and good sample standard deviation (with division by N - 1) as well as good MLE estimates (divde by N) in standalone functions.

That compiler explorer is super cool!  

In the past, we've found it hard to measure differences in these kinds of loop structures.  I know memory locality and vectorizability in the CPU are both critical, but I have roughly zero intuition on how to guess what's going to happen.

Almost all of our calculations are dominated by double precision to the point where integer arithmetic can be executed in parallel on the CPU for what seems like zero measurable cost.  

I also don't have good intuitions about whether wrapping some of this raw memory in `Eigen::Map` and letting Eigen sort out the loops would be more efficient.  Presumably it would be for matrices.

**Comment by danluu** [comment]
I submitted a small, naive, fix for the problems mentioned above since the change is trivial (add a couple of characters, move a couple lines around). It sounds like the better fix shouldn't be hard, but is a somewhat larger change.

**Comment by mcol** [comment]
This was originally fixed in #817, then it was overhauled in #1283 to use `Eigen::Map`, so I see no reason to keep it open.
